opensearch-operator/pkg/reconcilers/ismpolicy.go

```diff
@@ -0,0 +1,531 @@
+		exists, retErr = services.PolicyExists(r.ctx, r.osClient, r.instance.Spec.PolicyID)
+		if retErr != nil {
+			reason = "failed to get policy status from Opensearch API"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		if pointer.BoolDeref(r.updateStatus, true) {
+			retErr = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+					return err
+				}
+				r.instance.Status.ExistingISMPolicy = &exists
+				return r.Status().Update(r.ctx, r.instance)
+			})
+			if retErr != nil {
+				reason = fmt.Sprintf("failed to update status: %s", retErr)
+				r.recorder.Event(r.instance, "Warning", statusError, reason)
+				return
+			}
+		} else {
+			// Emit an event for unit testing assertion
+			r.recorder.Event(r.instance, "Normal", "UnitTest", fmt.Sprintf("exists is %t", exists))
+			return
+		}
+	}
+
+	// If ism policy is existing do nothing
+	if *r.instance.Status.ExistingISMPolicy {
+		reason = ismPolicyExists
+		return
+	}
+	ismpolicy, retErr := r.CreateISMPolicyRequest()
+	if retErr != nil {
+		reason = "failed to get create the ism policy request"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+
+	ismResponse, retErr := services.GetPolicy(r.ctx, r.osClient, r.instance.Spec.PolicyID)
+	if retErr != nil && retErr != services.ErrNotFound {
+		reason = "failed to get policy from Opensearch API"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+	if errors.Is(retErr, services.ErrNotFound) {
+		r.logger.V(1).Info(fmt.Sprintf("policy %s not found, creating.", r.instance.Spec.PolicyID))
+		retErr = services.CreateISMPolicy(r.ctx, r.osClient, *ismpolicy, r.instance.Spec.PolicyID)
+		if retErr != nil {
+			reason = "failed to create ism policy"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		r.recorder.Event(r.instance, "Normal", opensearchAPIUpdated, "policy created in opensearch")
+		return ctrl.Result{Requeue: true, RequeueAfter: 30 * time.Second}, retErr
+	}
+	if err != nil {
+		return
+	}
+	priterm := ismResponse.PrimaryTerm
+	seqno := ismResponse.SequenceNumber
+	// Reset
+	ismResponse.PrimaryTerm = nil
+	ismResponse.SequenceNumber = nil
+	shouldUpdate, retErr := services.ShouldUpdateISMPolicy(r.ctx, *ismpolicy, *ismResponse)
+	if retErr != nil {
+		reason = "failed to compare the policies"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+
+	if !shouldUpdate {
+		r.logger.V(1).Info(fmt.Sprintf("policy %s is in sync", r.instance.Spec.PolicyID))
+		return ctrl.Result{Requeue: true, RequeueAfter: 30 * time.Second}, retErr
+	}
+
+	if r.instance.Spec.PolicyID != ismResponse.PolicyID {
```

Comment:
The comparsion here doesn't make any sense as `ismResponse` is retrieved using `r.instance.Spec.PolicyId`.
You'll need to do it like it was done here (https://github.com/Opster/opensearch-k8s-operator/blob/main/opensearch-operator/pkg/reconcilers/indextemplate.go#L201) for the index templates by storing the policyId in the status.